### PR TITLE
Update/rollback in maintenance window

### LINKED ISSUE #553: modify the update flow to use maintenance window

### PR DESCRIPTION
**Pull request by yanganto**
Fix #553 
Introduce `--immediately` options to `kuutamo update/rollback`
```text
If immediately:
  do the current update/rollback flow
else:
  Shutdown kuutamod in a maintenance window (when waiting for shutdown, use `kuutamoctl maintenance-status` to show the current block and the maintenance setting)
  then do update/rollback
```

### CODE REVIEW

File: src/bin/kuutamoctl.rs [hunk 1]
```diff
@@ -42,27 +45,58 @@ async fn show_active_validator(kuutamo_client: &CommandClient, args: &Args) -> R
...
+            ..
+        }) => {
+            if minimum_length.is_some() && shutdown_at.is_some() {
+                Err(anyhow!("We can not guarantee minimum maintenance window for a specified shutdown block height"))
+            } else {
+                let r = kuutamo_client
+                    .maintenance_operation(minimum_length, shutdown_at, cancel, false)
+                    .await;
+                if r.is_ok() && wait {
+                    while UnixStream::connect(&args.control_socket).await.is_ok() {
```
Comment (Mic92) [inline src/bin/kuutamoctl.rs:73]: Is this is what is happening here?

```suggestion
                    // Wait for kuutamod to terminate
                    while UnixStream::connect(&args.control_socket).await.is_ok() {
```

File: src/commands/server.rs [hunk 2]
```diff
@@ -141,6 +159,39 @@ impl CommandServer {
...
+        let resp = match (
+            metrics.ok().and_then(|m| {
+                m.get("near_block_expected_shutdown")
+                    .and_then(|s| s.parse::<u64>().ok())
+            }),
+            final_block,
+            SHUTDOWN_WITH_NEARD.load(Ordering::Acquire),
+        ) {
+            (Some(expect), Ok(current), true) if expect > 0 => Response::new(Body::from(format!(
+                "{{\"status\": 200, \"message\": \"maintenance shutdown in {} blocks, current: {current:}, shutdwon at: {expect}\"}}",
```
Comment (Mic92) [inline src/commands/server.rs:176]:
```suggestion
                "{{\"status\": 200, \"message\": \"maintenance shutdown in {} blocks, current: {current:}, shutdown at: {expect}\"}}",
```

File: src/commands/server.rs [hunk 3]
```diff
@@ -141,6 +159,39 @@ impl CommandServer {
...
+                "{{\"status\": 200, \"message\": \"maintenance restart in {} blocks,  current: {current:}, restart at: {expect}\"}}",
+                expect - current
+            ))),
+            (Some(expect), Err(_), true) if expect > 0 => Response::new(Body::from(format!("{{\"status\": 200, \"message\": \"maintenance shutdown will be at {expect}\"}}"))),
+            (Some(expect), Err(_), false) if expect > 0 => Response::new(Body::from(format!("{{\"status\": 200, \"message\": \"maintenance restart will be at {expect}\"}}"))),
+            (_, Ok(current), true) => Response::new(Body::from(format!(
+                "{{\"status\": 200, \"message\": \"maintenance shutdown set, current: {current:}\"}}",
+            ))),
+            (_, _, false) => Response::new(Body::from("{\"status\": 200, \"message\": \"no maintenance setting now\"}")),
+            (_, Err(_), _) => server_error("fail to fetch current block from neard"),
```
[the remark this answers is not in the document]
Comment (yanganto) [reply src/commands/server.rs]: Using gateway timeout when we can not get the response of the current block from neard

**Comment by yanganto** [comment]
Thanks for reviewing.

File: src/supervisor.rs [hunk 4]
```diff
@@ -52,6 +55,8 @@ const CONSUL_ACQUIRE_LEADER_FREQUENCY: Duration = Duration::from_secs(1);
 const CONSUL_LEADER_TIMEOUT: Duration = Duration::from_secs(25);
 /// How often we query neard's `/status` endpoint
 const NEARD_STATUS_FREQUENCY: Duration = Duration::from_secs(1);
+/// Shutdown kuutamod if neard shutdwon as expected
```
Comment (Mic92) [inline src/supervisor.rs:58]:
```suggestion
/// Shutdown kuutamod if neard shutdown as expected
```

File: src/ipc.rs [hunk 5]
```diff
@@ -14,13 +14,17 @@ pub struct MaintenanceShutdownResponse {
...
 pub enum Request {
-    /// Schedule maintenance shutdown for a maintainace window of given size, book a block
-    /// height to shutdown at or cancel previous shutdon setting
-    /// + Channel where the supervisor will respond to once the request is finished
-    MaintenanceShutdown(
+    /// Schedule maintenance shutdown for
+    ///     * a maintainace window of given size,
+    ///     * book a block height to shutdown at,
+    ///     * cancel previous shutdon setting
+    ///     * shutdown kuutamod after nerd shutdown
```
Comment (Mic92) [inline src/ipc.rs:21]:
```suggestion
    ///     * shutdown kuutamod after neard shutdown
```

File: src/deploy/utils.rs [hunk 6]
```diff
@@ -22,3 +27,73 @@ pub fn timeout_ssh(host: &Host, command: &[&str], learn_known_host_key: bool) ->
...
+    }
+}
+
+/// Keep printing maintenance status before maintenance shutdown
+pub async fn handle_maintenance_shutdown(
+    host: &Host,
+    required_time_in_blocks: Option<u64>,
+) -> Result<()> {
+    let flag = AtomicBool::new(true);
+    let shutdwon_cmd = if let Some(blocks) = required_time_in_blocks {
```
Comment (Mic92) [inline src/deploy/utils.rs:73]:
```suggestion
    let shutdown_cmd = if let Some(blocks) = required_time_in_blocks {
```

**Comment by Mic92** [comment]
bors merge

**Comment by yanganto** [comment]
bors merge

**Pull request merged**